File: holochain/rna/rust/src/report.rs
//! Migration reporting and verification structures
//!
//! # RNA Metaphor: Transcription Reports
//!
//! Just as biologists track gene expression levels and transcription success,
//! these structures track what was successfully "transcribed" from one DNA to another.
//!
//! # Usage
//!
//! ```rust,ignore
//! use hc_rna::{MigrationReport, MigrationVerification};
//!
//! let mut report = MigrationReport::new("v1".to_string(), "v2".to_string());
//!
//! // Track successes and failures
//! report.record_success("Content");
//! report.record_failure("Path", Some("path-123".to_string()), "Invalid reference".to_string());
//!
//! // Complete the report
//! report.complete();
//!
//! println!("Migration success: {}", report.is_success());
//! ```

use hdk::prelude::*;
use std::collections::HashMap;

/// Migration report tracking success/failure of migrated items
///
/// This is the "transcription report" - documenting what was successfully
/// transcribed from the source DNA to the target DNA.
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MigrationReport {
    /// Schema version of the source DNA
    pub source_version: String,
    /// Schema version of the target DNA
    pub target_version: String,
    /// When migration started (ISO 8601 or Holochain timestamp)
    pub started_at: String,
    /// When migration completed (None if still running)
    pub completed_at: Option<String>,
    /// Per-entry-type migration counts
    pub entry_counts: HashMap<String, MigrationCounts>,
    /// List of errors encountered
    pub errors: Vec<MigrationError>,
    /// Post-migration verification results
    pub verification: MigrationVerification,
}

/// Counts for a single entry type
#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct MigrationCounts {
    /// Number of entries exported from source
    pub exported: u32,
    /// Number of entries transformed
    pub transformed: u32,
    /// Number of entries successfully imported
    pub imported: u32,
    /// Number of entries skipped (already exist)
    pub skipped: u32,
    /// Number of entries that failed to import
    pub failed: u32,
}

/// A single migration error
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MigrationError {
    /// The type of entry that failed
    pub entry_type: String,
    /// The ID of the specific entry (if available)
    pub entry_id: Option<String>,
    /// Which phase the error occurred in
    pub phase: MigrationPhase,
    /// Human-readable error message
    pub message: String,
}

/// Phase of migration where an error occurred
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum MigrationPhase {
    /// Error during export from source DNA
    Export,
    /// Error during schema transformation
    Transform,
    /// Error during import to target DNA
    Import,
    /// Error during verification
    Verify,
}

/// Verification results after migration
#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct MigrationVerification {
    /// Overall verification passed
    pub passed: bool,
    /// Per-entry-type count checks
    pub count_checks: HashMap<String, CountCheck>,
    /// All references resolve correctly
    pub reference_integrity: bool,
    /// Additional notes or warnings
    pub notes: Vec<String>,
}

/// Result of checking expected vs actual counts
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct CountCheck {
    /// Expected count (from source)
    pub expected: u32,
    /// Actual count (in target)
    pub actual: u32,
    /// Whether this check passed
    pub passed: bool,
}

impl MigrationReport {
    /// Create a new migration report
    ///
    /// # Arguments
    ///
    /// * `source_version` - Schema version of source DNA (e.g., "v1")
    /// * `target_version` - Schema version of target DNA (e.g., "v2")
    pub fn new(source_version: String, target_version: String) -> Self {
        let now = sys_time()
            .ok()
            .map(|t| format!("{:?}", t))
            .unwrap_or_else(|| chrono_now());

        Self {
            source_version,
            target_version,
            started_at: now,
            completed_at: None,
            entry_counts: HashMap::new(),
            errors: Vec::new(),
            verification: MigrationVerification::default(),
        }
    }

    /// Record a successful import
    pub fn record_success(&mut self, entry_type: &str) {
        let counts = self.entry_counts.entry(entry_type.to_string()).or_default();
        counts.imported += 1;
    }

    /// Record a skipped entry (already exists)
    pub fn record_skip(&mut self, entry_type: &str) {
        let counts = self.entry_counts.entry(entry_type.to_string()).or_default();
        counts.skipped += 1;
    }

    /// Record a failed import
    pub fn record_failure(&mut self, entry_type: &str, entry_id: Option<String>, message: String) {
        let counts = self.entry_counts.entry(entry_type.to_string()).or_default();
        counts.failed += 1;
        self.errors.push(MigrationError {
            entry_type: entry_type.to_string(),
            entry_id,
            phase: MigrationPhase::Import,
            message,
        });
    }

    /// Record an export count
    pub fn record_exported(&mut self, entry_type: &str, count: u32) {
        let counts = self.entry_counts.entry(entry_type.to_string()).or_default();
        counts.exported = count;
    }

    /// Record a transform count
    pub fn record_transformed(&mut self, entry_type: &str, count: u32) {
        let counts = self.entry_counts.entry(entry_type.to_string()).or_default();
        counts.transformed = count;
    }

    /// Add an error for any phase
    pub fn add_error(&mut self, entry_type: &str, phase: MigrationPhase, message: String) {
        self.errors.push(MigrationError {
            entry_type: entry_type.to_string(),
            entry_id: None,
            phase,
            message,
        });
    }

    /// Mark the migration as complete
    pub fn complete(&mut self) {
        let now = sys_time()
            .ok()
            .map(|t| format!("{:?}", t))
            .unwrap_or_else(|| chrono_now());
        self.completed_at = Some(now);
    }

    /// Check if migration was successful (no failures, verification passed)
    pub fn is_success(&self) -> bool {
        let no_failures = self
            .entry_counts
            .values()
            .all(|c| c.failed == 0);
        no_failures && self.verification.passed
    }

    /// Get total count of successfully imported entries
    pub fn total_imported(&self) -> u32 {
        self.entry_counts.values().map(|c| c.imported).sum()
    }

    /// Get total count of failed entries
    pub fn total_failed(&self) -> u32 {
        self.entry_counts.values().map(|c| c.failed).sum()
    }

    /// Get total count of skipped entries
    pub fn total_skipped(&self) -> u32 {
        self.entry_counts.values().map(|c| c.skipped).sum()
    }
}

impl MigrationVerification {
    /// Create a new verification with all checks passing
    pub fn passed() -> Self {
        Self {
            passed: true,
            count_checks: HashMap::new(),
            reference_integrity: true,
            notes: Vec::new(),
        }
    }

    /// Add a count check result
    pub fn add_count_check(&mut self, entry_type: &str, expected: u32, actual: u32) {
        let passed = actual >= expected;
        self.count_checks.insert(
            entry_type.to_string(),
            CountCheck {
                expected,
                actual,
                passed,
            },
        );
        // Update overall passed status
        self.passed = self.passed && passed;
    }

    /// Add a note
    pub fn add_note(&mut self, note: String) {
        self.notes.push(note);
    }

    /// Mark reference integrity check result
    pub fn set_reference_integrity(&mut self, passed: bool) {
        self.reference_integrity = passed;
        self.passed = self.passed && passed;
    }
}

/// Fallback timestamp when sys_time() not available (outside wasm)
fn chrono_now() -> String {
    "unknown".to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_migration_report_basic() {
        let mut report = MigrationReport::new("v1".to_string(), "v2".to_string());

        report.record_success("Content");
        report.record_success("Content");
        report.record_skip("Content");
        report.record_failure("Path", Some("path-1".to_string()), "Bad ref".to_string());

        assert_eq!(report.entry_counts.get("Content").unwrap().imported, 2);
        assert_eq!(report.entry_counts.get("Content").unwrap().skipped, 1);
        assert_eq!(report.entry_counts.get("Path").unwrap().failed, 1);
        assert_eq!(report.errors.len(), 1);
    }

    #[test]
    fn test_verification() {
        let mut verification = MigrationVerification::passed();

        verification.add_count_check("Content", 10, 10);
        assert!(verification.passed);

        verification.add_count_check("Path", 5, 3);
        assert!(!verification.passed);
    }
}

File: holochain/rna/typescript/src/report.ts
/**
 * Migration Report Types
 *
 * # RNA Metaphor: Transcription Reports
 *
 * These types match the Rust report structures for serialization
 * between TypeScript and the Holochain zome.
 */

/**
 * Complete migration report
 */
export interface MigrationReport {
  /** Schema version of source DNA */
  source_version: string;
  /** Schema version of target DNA */
  target_version: string;
  /** When migration started */
  started_at: string;
  /** When migration completed (null if still running) */
  completed_at: string | null;
  /** Per-entry-type migration counts */
  entry_counts: Record<string, MigrationCounts>;
  /** List of errors encountered */
  errors: MigrationError[];
  /** Post-migration verification results */
  verification: MigrationVerification;
}

/**
 * Counts for a single entry type
 */
export interface MigrationCounts {
  /** Number exported from source */
  exported: number;
  /** Number transformed */
  transformed: number;
  /** Number successfully imported */
  imported: number;
  /** Number skipped (already exist) */
  skipped: number;
  /** Number failed to import */
  failed: number;
}

/**
 * A single migration error
 */
export interface MigrationError {
  /** Entry type that failed */
  entry_type: string;
  /** Specific entry ID if available */
  entry_id: string | null;
  /** Phase where error occurred */
  phase: MigrationPhase;
  /** Error message */
  message: string;
}

/**
 * Migration phase for error tracking
 */
export type MigrationPhase = 'Export' | 'Transform' | 'Import' | 'Verify';

/**
 * Verification results
 */
export interface MigrationVerification {
  /** Overall verification passed */
  passed: boolean;
  /** Per-type count checks */
  count_checks: Record<string, CountCheck>;
  /** Reference integrity passed */
  reference_integrity: boolean;
  /** Additional notes */
  notes: string[];
}

/**
 * Single count check result
 */
export interface CountCheck {
  /** Expected count from source */
  expected: number;
  /** Actual count in target */
  actual: number;
  /** Whether this check passed */
  passed: boolean;
}

/**
 * Create an empty migration report
 */
export function createReport(
  sourceVersion: string,
  targetVersion: string
): MigrationReport {
  return {
    source_version: sourceVersion,
    target_version: targetVersion,
    started_at: new Date().toISOString(),
    completed_at: null,
    entry_counts: {},
    errors: [],
    verification: {
      passed: false,
      count_checks: {},
      reference_integrity: false,
      notes: [],
    },
  };
}

/**
 * Record a successful import in the report
 */
export function recordSuccess(report: MigrationReport, entryType: string): void {
  if (!report.entry_counts[entryType]) {
    report.entry_counts[entryType] = {
      exported: 0,
      transformed: 0,
      imported: 0,
      skipped: 0,
      failed: 0,
    };
  }
  report.entry_counts[entryType].imported++;
}

/**
 * Record a skipped entry (already exists)
 */
export function recordSkip(report: MigrationReport, entryType: string): void {
  if (!report.entry_counts[entryType]) {
    report.entry_counts[entryType] = {
      exported: 0,
      transformed: 0,
      imported: 0,
      skipped: 0,
      failed: 0,
    };
  }
  report.entry_counts[entryType].skipped++;
}

/**
 * Record a failed import
 */
export function recordFailure(
  report: MigrationReport,
  entryType: string,
  entryId: string | null,
  message: string
): void {
  if (!report.entry_counts[entryType]) {
    report.entry_counts[entryType] = {
      exported: 0,
      transformed: 0,
      imported: 0,
      skipped: 0,
      failed: 0,
    };
  }
  report.entry_counts[entryType].failed++;
  report.errors.push({
    entry_type: entryType,
    entry_id: entryId,
    phase: 'Import',
    message,
  });
}

/**
 * Mark the report as complete
 */
export function completeReport(report: MigrationReport): void {
  report.completed_at = new Date().toISOString();
}

/**
 * Check if migration was successful
 */
export function isSuccess(report: MigrationReport): boolean {
  const noFailures = Object.values(report.entry_counts).every(
    (c) => c.failed === 0
  );
  return noFailures && report.verification.passed;
}

/**
 * Get total imported count
 */
export function totalImported(report: MigrationReport): number {
  return Object.values(report.entry_counts).reduce(
    (sum, c) => sum + c.imported,
    0
  );
}

/**
 * Get total failed count
 */
export function totalFailed(report: MigrationReport): number {
  return Object.values(report.entry_counts).reduce(
    (sum, c) => sum + c.failed,
    0
  );
}

/**
 * Format report for console output
 */
export function formatReport(report: MigrationReport): string {
  const lines: string[] = [
    '='.repeat(60),
    '  Migration Report',
    '='.repeat(60),
    `Source: ${report.source_version} -> Target: ${report.target_version}`,
    `Started: ${report.started_at}`,
    `Completed: ${report.completed_at || 'In progress'}`,
    '',
    'Entry Counts:',
  ];

  for (const [type, counts] of Object.entries(report.entry_counts)) {
    lines.push(
      `  ${type}: ${counts.imported} imported, ${counts.skipped} skipped, ${counts.failed} failed`
    );
  }

  if (report.errors.length > 0) {
    lines.push('', 'Errors:');
    for (const err of report.errors.slice(0, 10)) {
      const id = err.entry_id ? ` (${err.entry_id})` : '';
      lines.push(`  [${err.phase}] ${err.entry_type}${id}: ${err.message}`);
    }
    if (report.errors.length > 10) {
      lines.push(`  ... and ${report.errors.length - 10} more`);
    }
  }

  lines.push('', `Verification: ${report.verification.passed ? 'PASSED' : 'FAILED'}`);

  if (report.verification.notes.length > 0) {
    for (const note of report.verification.notes) {
      lines.push(`  - ${note}`);
    }
  }

  lines.push('='.repeat(60));

  return lines.join('\n');
}
